### PR TITLE
Initial documentation for index templates V2

### PR DESCRIPTION
**Pull request by dakrone**
This commit adds the initial document for version two index templates. Since these are intended to
be used in favor of V1 index templates, this re-uses the anchors for index templates to point to the
V2 APIs, renaming the V1 template docs' anchors (this was a suggestion from the docs team).

There is more documentation that can be written, but this is a start.

Relates to #53101

### CODE REVIEW

File: docs/reference/indices/templates.asciidoc [hunk 1]
```diff
@@ -1,9 +1,21 @@
...
 
+This documentation is about V1 (version 1) index templates, are deprecated and will be replaced by
+V2 templates. For V2 templates please see the <<indices-templates-v1,V1 template documentation>>.
+
+[NOTE]
+====
+In {es} 7.8 the `prefer_v2_templates` querystring parameter was introduced to the index, update,
+bulk, and create index APIs to allow configuring whether to favor V2 or V1 index templates. In 7.x
+the default value for this parameter is `false` meaning that the V1 index templates take precedence.
+In 8.x the default value for the parameter is `true` meaning that <<indices-templates,V2 index
```
Comment (debadair) [inline docs/reference/indices/templates.asciidoc:15]: This xref is causing the build failure. You either need to change the anchor in index-templates.asciidoc to [[indices-templates]], or update this xref. 

I think I'd stick with the indices-templates anchor that was used for v1, as there are likely external links to the topic. I know the plural indices is weird, but that pattern is consistent through the rest of the index APIs.

File: docs/reference/indices/templates.asciidoc [hunk 2]
```diff
@@ -1,9 +1,21 @@
-[[indices-templates]]
+[[indices-templates-v1]]
 === Put index template API
 ++++
 <titleabbrev>Put index template</titleabbrev>
 ++++
 
+This documentation is about V1 (version 1) index templates, are deprecated and will be replaced by
+V2 templates. For V2 templates please see the <<indices-templates-v1,V1 template documentation>>.
```
Comment (debadair) [inline docs/reference/indices/templates.asciidoc:8]:
```suggestion
V2 templates. For information about  V2 templates, see <<indices-templates>>.
```

**Review by debadair — CHANGES_REQUESTED** [review]
Left a couple comments related to the build errror.

File: docs/reference/indices/component-templates.asciidoc [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+<<indices-aliases,aliases>> that you can automatically apply when creating new indices. On their own
+component templates are not matched or applied to an index, instead they are used in the
+`composed_of` definition of an <<index-templates-v2,index template>> to make up the final index
+configuration.
+// end::component-template-def[]
+
+Component templates are only used during index creation.
+Changes to component templates do not affect existing indices.
+Settings and mappings specified in <<indices-create-index, create index>> API requests
+override any settings or mappings specified in a component template.
```
Comment (andreidan) [inline docs/reference/indices/component-templates.asciidoc:70]: Should we mention this in the index template (`<<index-templates-v2,index templates>>`) documentation? (as component templates on their own don't interact with indices in any way)

File: docs/reference/indices/templates.asciidoc [hunk 4]
```diff
@@ -1,9 +1,21 @@
-[[indices-templates]]
+[[indices-templates-v1]]
 === Put index template API
 ++++
 <titleabbrev>Put index template</titleabbrev>
 ++++
 
+This documentation is about V1 (version 1) index templates, are deprecated and will be replaced by
```
Comment (andreidan) [inline docs/reference/indices/templates.asciidoc:7]: is a "which" missing here? ie. ```This documentation is about V1 (version 1) index templates, *which* are deprecated and will be replaced```

**Review by andreidan — APPROVED** [review]
LGTM thanks for adding the docs Lee. I've left 2 minor suggestions, but this is overall great

File: docs/reference/indices/templates.asciidoc (hunk 4 above)
Comment (dakrone) [reply docs/reference/indices/templates.asciidoc:7]: Yes it is missing a "which", thanks!

File: docs/reference/indices/component-templates.asciidoc (hunk 3 above)
Comment (dakrone) [reply docs/reference/indices/component-templates.asciidoc:70]: I do have a mention of this at the top of the index template docs, but I can make it a bit clearer.

File: docs/reference/indices/templates.asciidoc (hunk 1 above)
Comment (dakrone) [reply docs/reference/indices/templates.asciidoc:15]: Sounds good, thanks for taking a look @debadair!

**Review by probakowski — APPROVED** [review]
LGTM 👍

**Review by debadair — APPROVED** [review]
Left a few minor comments/suggestions.

**Comment by dakrone** [comment]
Thanks for the reviews all!

**Pull request merged**